src/UCTSearch.cpp

```diff
@@ -49,12 +50,71 @@ void UCTSearch::update_root() {
     // Definition of m_playouts is playouts per search call.
     // So reset this count now.
     m_playouts = 0;
-    // See if the current state of the game is a direct succesor of
-    // our stored tree's root node. If so, make that the new root.
-    UCTNode::find_new_root(m_root, m_rootstate, std::move(m_last_rootstate));
-    // If the above succeeded, we'll have a search tree already. See
-    // how big it is.
+
+    bool found = false;
+    #ifndef NDEBUG
+    auto start_nodes = m_root->count_nodes();
+    std::string replay;
+    #endif
+
+    // Check if new rootstate is a possible descendant of our stored rootstate.
+    if (m_last_rootstate
+            && m_rootstate.get_komi() == m_last_rootstate->get_komi()
+            && m_rootstate.get_movenum() >= m_last_rootstate->get_movenum()) {
+        auto depth = m_rootstate.get_movenum() - m_last_rootstate->get_movenum();
+        auto test = std::make_unique<GameState>(m_rootstate);
+        for (auto i = size_t{0}; i < depth; i++) {
+            test->undo_move();
+        }
+
+        // Try to replay moves advancing m_root
+        for (auto i = size_t{0}; i < depth; i++) {
+            // Verify m_rootstate and m_last_rootstate have same history.
+            if (m_last_rootstate->board.get_ko_hash()
+                    != test->board.get_ko_hash()) {
+                break;
+            }
+
+            test->forward_move();
+            auto move = test->get_last_move();
+            m_root = m_root->find_child(move);
+            if (!m_root->has_children()) {
```

Comment:
You will crash here if it's a nullptr.

Reply:
I think you always hit an expanded node first but it's easy to add.